contracts/RegistrarController.sol

```diff
@@ -0,0 +1,161 @@
+     * @param _domainPrice Init domain price
+     */
+    function initialize(address _owner, uint _domainPrice) initializer public {
+        __Ownable_init(_owner);
+        _getMainStorage().domainPrice = _domainPrice;
+        emit DomainPriceChanged(_domainPrice);
+    }
+
+    /**
+     * Return controller of domain
+     * @param domainName Name of domain to check
+     */
+    function getDomainController(string memory domainName) external view returns (address) {
+        return _getMainStorage().domainRecords[domainName];
+    }
+
+    function domainPrice() external view returns (uint256) {
+        return _getMainStorage().domainPrice;
+    }
+
+    function domainRewards(address addressToCheck) public view returns (uint) {
+        return _getMainStorage().domainRewards[addressToCheck];
+    }
+
+    /**
+     * Register new domain
+     * @param domainName Name of domain to register
+     */
+    function registerDomain(string memory domainName) payable external {
+        MainStorage storage $ = _getMainStorage();
+
+        require(msg.value >= $.domainPrice, "Ether value is lower than price.");
+        require($.domainRecords[domainName] == address(0), "Domain has been purchased by someone before.");
+
+        strings.slice memory s = domainName.toSlice();
+        strings.slice memory delim = ".".toSlice();
+        uint parts = s.count(delim);
+
+        uint restValue = $.domainPrice;
+
+        string memory localDomainName = "";
+        
+        for(uint i = 0; i < parts; i++) {
+            if (i > 0) {
+                localDomainName = string.concat(s.rsplit(delim).toString(), ".", localDomainName); // Join all domain parts
+            } else {
+                localDomainName = s.rsplit(delim).toString(); // or take last part of domain, if it first iteration
+            }
+            
+            
+            address domainOwner = $.domainRecords[localDomainName];
+            require(domainOwner != address(0), "Not all domain levels has been registred.");
+
+            if (domainOwner != address(0)) {
+                uint reward = $.domainPrice / parts;
+                $.domainRewards[domainOwner] += reward;
+
+                restValue -= reward;
+            }
+        }
+
+        $.domainRewards[owner()] += restValue;
+
+        $.domainRecords[domainName] = msg.sender;
+        emit DomainPurchase(msg.sender, domainName, block.timestamp);
+    }
+
+    /**
+     * Set new price for domains (Only for owner)
+     * @param newPrice New price for domains
+     */
+    function setDomainPrice(uint newPrice) external onlyOwner {
+        MainStorage storage $ = _getMainStorage();
+        $.domainPrice = newPrice;
+        emit DomainPriceChanged($.domainPrice);
+    }
+
+    /**
+     * Withdraws all ethers rewards from contract to owner
+     * @param to Rewards receiver address
+     */
+    function withdrawAllEther(address payable to) external onlyOwner {
```

Comment:
It is better to realize this function without only owner 
Right now, the owner spends a gas fee to transfer ETH to each receiver. 

It will be better to give the possibility of withdrawal to anyone who has some balance.